gene/delete.go

```diff
@@ -0,0 +1,338 @@
+package gene
+
+import (
+	"errors"
+	"github.com/vertgenlab/gonomics/dna"
+)
+
+var (
+	ErrNoStopFound = errors.New("frameshift was unable to find a stop before the end of the gene")
+)
+
+//TODO EffectPrediction
+// Deletion removes bases from the Gene, predicts the effect, and updates the Gene struct to reflect the change.
+// genomeStartPos should be the base directly BEFORE the deleted bases. genomeEndPos should be the base directly AFTER the deleted bases.
+// All positions should be given as base-zero genomic coordinates.
+func Deletion(g *Gene, genomeStartPos int, genomeEndPos int) (EffectPrediction, error) {
+	var answer EffectPrediction
+	var err error
+
+	genomeIndexStartPos, genomeIndexEndPos, err := deletionPreRunChecks(g, genomeStartPos, genomeEndPos)
+	if err != nil {
+		return EffectPrediction{}, err
+	}
+
+	// Save CDS start and stop for later
+	origCdsStart := g.cdsStarts[0]
+	origCdsEnd := g.cdsEnds[len(g.cdsEnds)-1]
+
+	// Delete from Genome Sequence
+	copy(g.genomeSeq[genomeIndexStartPos+1:], g.genomeSeq[genomeIndexEndPos:])
+	g.genomeSeq = g.genomeSeq[:len(g.genomeSeq)-(genomeIndexEndPos-(genomeIndexStartPos+1))]
+
+	// Update startPos if necessary
+	if genomeIndexStartPos == -1 {
+		g.startPos += genomeIndexEndPos - (genomeIndexStartPos + 1)
+	}
+
+	// Update CDS starts and ends and calculate how many cDNA bases have been deleted
+	var cdsIndexesToDelete []int
+	var deletedCodingBases int
+	var codingDelStart, codingDelEnd int = -1, -1
+	deletionLen := genomeIndexEndPos - (genomeIndexStartPos + 1)
+	for i := 0; i < len(g.cdsStarts); i++ {
+
+		switch {
+		// if cds is before deletion
+		case genomeIndexStartPos >= g.cdsEnds[i]:
+			codingDelStart = int(g.featureArray[g.cdsEnds[i]])
+
+		// If cds exon is fully deleted
+		case genomeIndexStartPos < g.cdsStarts[i] && genomeIndexEndPos > g.cdsEnds[i]:
```

Comment:
Should the > be >=?  I am thinking that you need the < because the deletion is left-open and the cdsStart is left-closed?

Reply:
since the index positions were fully open and the cds positions were fully closed this logic was correct, however it has been changed since making the index positions left closed right open